# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/workers/loader/CacheLoadHandler.cpp
# Commit: 1f7eacb7b39f
# Full Hash: 1f7eacb7b39f76128f7c969020f2d6ba0242225f
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2022-11-22 09:46:06
# Description:
#   Bug 1801616 - Cancellation should abort network and cache loads if it has already completed;  r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D162557
# ==============================================================================

diff -r aaffb4cfb02e -r 1f7eacb7b39f dom/workers/loader/CacheLoadHandler.cpp
--- a/dom/workers/loader/CacheLoadHandler.cpp	Mon Nov 21 15:03:04 2022 -0500
+++ b/dom/workers/loader/CacheLoadHandler.cpp	Mon Nov 21 18:23:20 2022 +0000
@@ -51,7 +51,9 @@
                                            JS::Handle<JS::Value> aValue,
                                            ErrorResult& aRv) {
   AssertIsOnMainThread();
-  MOZ_ASSERT(!mRequestHandle->IsEmpty());
+  if (mRequestHandle->IsEmpty()) {
+    return;
+  }
   WorkerLoadContext* loadContext = mRequestHandle->GetContext();
 
   // May already have been canceled by CacheLoadHandler::Fail from
@@ -72,7 +74,9 @@
                                            JS::Handle<JS::Value> aValue,
                                            ErrorResult& aRv) {
   AssertIsOnMainThread();
-  MOZ_ASSERT(!mRequestHandle->IsEmpty());
+  if (mRequestHandle->IsEmpty()) {
+    return;
+  }
   WorkerLoadContext* loadContext = mRequestHandle->GetContext();
 
   // May already have been canceled by CacheLoadHandler::Fail from
@@ -253,8 +257,6 @@
 
 void CacheLoadHandler::Fail(nsresult aRv) {
   AssertIsOnMainThread();
-  MOZ_ASSERT(!mRequestHandle->IsEmpty());
-  WorkerLoadContext* loadContext = mRequestHandle->GetContext();
   MOZ_ASSERT(NS_FAILED(aRv));
 
   if (mFailed) {
@@ -264,11 +266,17 @@
   mFailed = true;
 
   if (mPump) {
-    MOZ_ASSERT(loadContext->mCacheStatus ==
-               WorkerLoadContext::ReadingFromCache);
+    MOZ_ASSERT_IF(!mRequestHandle->IsEmpty(),
+                  mRequestHandle->GetContext()->mCacheStatus ==
+                      WorkerLoadContext::ReadingFromCache);
     mPump->Cancel(aRv);
     mPump = nullptr;
   }
+  if (mRequestHandle->IsEmpty()) {
+    return;
+  }
+
+  WorkerLoadContext* loadContext = mRequestHandle->GetContext();
 
   loadContext->mCacheStatus = WorkerLoadContext::Cancel;
 
@@ -484,7 +492,9 @@
                                    uint32_t aStringLen,
                                    const uint8_t* aString) {
   AssertIsOnMainThread();
-  MOZ_ASSERT(!mRequestHandle->IsEmpty());
+  if (mRequestHandle->IsEmpty()) {
+    return NS_OK;
+  }
   WorkerLoadContext* loadContext = mRequestHandle->GetContext();
 
   mPump = nullptr;
@@ -515,7 +525,9 @@
     const nsACString& aCSPReportOnlyHeaderValue,
     const nsACString& aReferrerPolicyHeaderValue) {
   AssertIsOnMainThread();
-  MOZ_ASSERT(!mRequestHandle->IsEmpty());
+  if (mRequestHandle->IsEmpty()) {
+    return NS_OK;
+  }
   WorkerLoadContext* loadContext = mRequestHandle->GetContext();
 
   MOZ_ASSERT(loadContext->mCacheStatus == WorkerLoadContext::Cached);